feat: FIT-587: FSM Architecture setup

bmartel opened the pull request:
This pull request introduces a new Finite State Machine (FSM) framework to the Label Studio codebase. The FSM system is designed for high-performance, time-series optimized state management using UUID7, and provides a generic, extensible infrastructure for tracking entity states and transitions. The changes include core model definitions, registry and transition systems, documentation, and Django app configuration. Additionally, the FSM framework is integrated into the main application settings with initial configuration.

**FSM Framework Introduction and Integration**

* Added the `fsm` app to the `INSTALLED_APPS` list in `base.py`, integrating the FSM framework into Label Studio's Django settings.
* Introduced FSM cache TTL configuration (`FSM_CACHE_TTL`) for controlling state cache duration.

**Core FSM Implementation**

* Implemented the core FSM model (`BaseState`) in `fsm/models.py`, featuring UUID7 primary keys for natural time ordering, audit trail fields, transition metadata, and optimized query methods for state history and time-range queries.
* Added initial state choices module (`fsm/state_choices.py`) for defining available states per entity type.

**Framework Documentation and App Structure**

* Added extensive documentation in `fsm/README.md`, detailing architecture, usage patterns, extension points, and performance characteristics of the FSM system.
* Created Django app configuration (`fsm/apps.py`) and package initialization (`fsm/__init__.py`) for the FSM framework. [[1]](diffhunk://#diff-b0bbdfca422af48284839dab1ab9ccb553fae0b4b4d3a2d0b3a8f1d68f103757R1-R13) [[2]](diffhunk://#diff-f699e3a7ad883bd25aaf854d487e2c1035d2c0b729a6eab3969f71defd15ab86R1-R12)
* Established migration and test package structure for future development (`fsm/migrations/__init__.py`, `fsm/tests/__init__.py`). [[1]](diffhunk://#diff-e46226c21c60230f56b6eafe1e0854fd24d92576a3cda5d10314d0b6044bcb5fR1) [[2]](diffhunk://#diff-56c3cc9bf9cd1c73e27ce972b2b564e16af7f42507d93516069de78ce50ac981R1)

bmartel committed:
feat: FIT-587: FSM Architecture setup

bmartel committed:
include organization on base state model

bmartel committed:
refactoring core definitions

bmartel committed:
define core entity states in full

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
fixed imports

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
fix imports

bmartel committed:
fix imports

bmartel committed:
fix fsm tests

bmartel committed:
fix fsm tests

bmartel committed:
fix linting errors

bmartel committed:
fix fsm concurrency tests

bmartel committed:
fix remaining fsm tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
updating doc strings

bmartel committed:
fixing registry

bmartel committed:
fixing registry

bmartel committed:
fixing tests

bmartel committed:
updating README

bmartel commented on label_studio/fsm/README.md:
````diff
@@ -0,0 +1,229 @@
+POST /api/fsm/{entity_type}/{entity_id}/transition/ # Execute transition
+```
+
+Extend the base viewset for your application:
````
```suggestion
Extend the base viewset
```

bmartel committed:
Apply suggestion from @bmartel

bmartel committed:
avoiding import cycle due to types

bmartel committed:
fix imports

bmartel committed:
fix imports

bmartel committed:
fix typing and decorator definition for state model

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
removing unused code

bmartel committed:
moving settings from extensions to settings file

bmartel committed:
moving the API integration FSM out of this PR

bmartel committed:
updating readme

bmartel committed:
removing unused code

bmartel committed:
renaming transition decorator for consistency

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
removing unused code

bmartel committed:
removing serializers from this PR, will be included in a different one

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

mcanu commented on label_studio/fsm/tests/test_api_usage_examples.py:
```diff
@@ -0,0 +1,792 @@
+            self.assertTrue(api_response['success'])
+            self.assertEqual(api_response['data']['new_state'], 'ASSIGNED')
+            self.assertEqual(api_response['data']['assignment_details']['assignee_id'], 123)
+            self.assertEqual(api_response['data']['assignment_details']['priority'], 'high')
```
To stay consistent with the rest of the tests lets use native `assert`

mcanu commented on label_studio/fsm/tests/test_api_usage_examples.py:
Lets rename this to test_api.py, the `APIIntegrationExampleTests` is already a logic separation. Or a `test_api.example.py` or `test_api_example.py` can also work.

We could also create a base class for new endpoints tests to inherit and have base assertions, but out of scope for this first approach.

mcanu commented on label_studio/fsm/tests/test_edge_cases_error_handling.py:
`test_error_handling.py`

mcanu commented on label_studio/fsm/tests/test_uuid7_utils.py:
`test_utils.py`

mcanu commented on label_studio/fsm/apps.py:
```diff
@@ -0,0 +1,46 @@
+                logger.info('FSM signal handlers registered')
+
+        except Exception as e:
+            logger.error(f'Failed to set up FSM signals: {e}')
```
Is this doing something?

bmartel replied:
It used to but I refactored a bunch of stuff as I didn't like the way this method of extending worked, I removed all of these in favour of just standard python OOP and decorators, and signals are not being used at this time, but even if they were, I'd prefer it if it was simply just a matter of standard Django signal usage. (tbh that is all this was doing, just in a more glorified manner 😆)

mcanu commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,434 @@
+    def mark_initialized(self):
+        """Mark the registry as initialized."""
+        self._initialized = True
+        logger.info(f'State model registry initialized with {len(self._models)} models')
```
Is this being used?

mcanu replied:
We can remove it to simplify

mcanu commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,434 @@
+    """
+
+    def __init__(self):
+        self._transitions: Dict[str, Dict[str, Type['BaseTransition']]] = {}
```
```suggestion
        self._transitions: Dict[str, Dict[str, 'BaseTransition']] = {}
```
I think you can do this like this, please check

mcanu replied:
No need to use `Type`

mcanu replied:
Same for the rest of the `Type` usage

mcanu commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,434 @@
+            for i, char in enumerate(class_name):
+                if char.isupper() and i > 0:
+                    name += '_'
+                name += char.lower()
```
I would make name required, just to be explicit about it and make it easier to search when troubleshooting.
Example, I see in the db a `start_task` transition and I want to search where its defined, I prefer not to remember this "magic" name creation.

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+        # Try cache first
+        cached_state = cache.get(cache_key)
+        if cached_state is not None:
+            logger.debug(f'Cache hit for {entity._meta.label_lower} {entity.pk}: {cached_state}')
```
Lets do this info and try to make these logs as structured as possible. I'd love to get some Grafana dashboard going with Cache Hit/Miss rates.

mcanu replied:
Also log the miss.

We can revisit this if we see we are flooded with logs.

hakan458 replied:
Yeah that might be a lot of logs 😆

bmartel replied:
I can structure this and move to level info 👍

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+                .order_by('-id')  # UUID7 natural ordering
+                .values_list('state', flat=True)
+                .first()
+            )
```
lets use `StateModel.get_current_state` here

hakan458 replied:
+1 , wonder if there is a reason why it wasnt used

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+            return None
+
+        entity_field = f'{entity._meta.model_name}'
+        return state_model.objects.filter(**{entity_field: entity}).order_by('-id').first()
```
Same here, lets leave the access to the StateModel to itself.

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+        logger.info(
+            f'Transitioning {entity._meta.label_lower} {entity.pk}: '
+            f'{current_state} → {new_state} (transition: {transition_name})'
+        )
```
Same here with structured logging. I can see a dashboard per Model state so we transactions being used. We can probably use the db for that, but I don't think we have any dashboards coming straight from the database.

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+                # Get denormalized fields from the state model itself
+                denormalized_fields = {}
+                if hasattr(state_model, 'get_denormalized_fields'):
+                    denormalized_fields = state_model.get_denormalized_fields(entity)
```
lets make `get_denormalized_fields` a default method in the base model and return an empty dict

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+            return []
+
+        entity_field = f'{entity._meta.model_name}'
+        return list(state_model.objects.filter(**{entity_field: entity}).order_by('-id')[:limit])
```
We already have this method in the State model. Also I wouldn't cast it to a list immediately, that would execute a query and not make it lazy in case it isn't used downstream

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+        if not state_model:
+            return []
+
+        return list(state_model.get_states_in_range(entity, start_time, end_time or datetime.now()))
```
Same here with list

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+    @classmethod
+    def warm_cache(cls, entities: List[Model]):
+        """
+        invalidate_cacheWarm cache with current states for a list of entities.
```
```suggestion
        Warm cache with current states for a list of entities.
```

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+        logger.debug(f'Invalidated cache for {entity._meta.label_lower} {entity.pk}')
+
+    @classmethod
+    def warm_cache(cls, entities: List[Model]):
```
Love this thinking

mcanu commented on label_studio/fsm/transition_utils.py:
```diff
@@ -0,0 +1,316 @@
+    return transition_registry.get_transitions_for_entity(entity_name)
+
+
+def get_valid_transitions(entity: Model, user=None, validate: bool = True) -> Dict[str, Type[BaseTransition]]:
```
Do we need the `validate` argument? We have the `get_available_transitions` for this.

bmartel replied:
Good catch, this is another refactor left-over, need to clean these up and consolidate.

bmartel replied:
We do need a validate argument as there are two modes of retrieval, one which gets available states from my current (validate=True) and the default is all available states for the entity (validate=False). I removed this extra function and consolidated the approach.

mcanu commented on label_studio/fsm/transition_utils.py:
```diff
@@ -0,0 +1,316 @@
+    )
+
+
+class TransitionBuilder:
```
Is this something we see ourselves using? I'd limit the number of ways we can create/execute transactions so the code is more maintainable.

bmartel replied:
Yea I personally don't see it, will remove it as honestly this is a more Java-esque pattern, not at all pythonic. Another concept that is leftover from all of the dabbling and thought process.

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+                f'State transition failed for {entity._meta.label_lower} {entity.pk}: '
+                f'{current_state} → {new_state}: {e}'
+            )
+            raise StateManagerError(f'Failed to transition state: {e}') from e
```
We aren't returning False if it failed, we are raising an error. Might be okay to raise an error and not return anything from this function. We'll need to update `BaseTransition.execute` that expects a boolean.

bmartel replied:
Good catch, I was pondering a concept to make the execution log or raise (reason being that in some circumstances, for example if we use this to backfill states, we wouldn't want it to fail hard, just log and move on). So I think in that regard we may want to tool the behaviour accordingly, and in the case of log only mode it should return whether it succeeded.

mcanu reviewed (commented):
Amazing work!! 😍

I'm not sure when I need to do a declarative transaction or not, lets leave that clear in the Readme. Or even limit to always wanting to create declarative transactions.

niklub commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,385 @@
+
+                # Update cache with new state
+                cache_key = cls.get_cache_key(entity)
+                cache.set(cache_key, new_state, cls.CACHE_TTL)
```
I wonder whether it's safer to write cache in the callback upon a successful commit, like
```
    transaction.on_commit(
        lambda: cache.set(cache_key, new_state, cls.CACHE_TTL)
    )
```

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Update label_studio/fsm/state_manager.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
updating based on feedback about test assertions/file names

bmartel committed:
updating based on feedback about test file names

bmartel committed:
consolidate transition util get_valid_states to being the same function get_available_states

bmartel committed:
improving transition handling

bmartel committed:
improving transition handling by consolidating everything to use a single entry point

bmartel committed:
ensure write through cache sets on_commit of the transaction

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
type fixes

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17446383898)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
reverting poetry.lock to develop to regenerate it on CI

bmartel committed:
Merge branch 'fb-fit-587-fsm' of github.com:HumanSignal/label-studio into fb-fit-587-fsm

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17446529166)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446529166

hakan458 commented on label_studio/core/settings/base.py:
```diff
@@ -891,3 +892,7 @@ def collect_versions_dummy(**kwargs):
+
+# Base FSM (Finite State Machine) Configuration for Label Studio
+FSM_CACHE_TTL = 300  # Cache TTL in seconds (5 minutes)
+FSM_AUTO_CREATE_STATES = False
```
Dont see any usages of this here or in LSE , is this future planned work?

bmartel replied:
Yea I took this out entirely, or so I thought 😆. Nice catch

hakan458 commented on label_studio/fsm/models.py:
```diff
@@ -0,0 +1,168 @@
+
+    def __str__(self):
+        entity_id = getattr(self, f'{self._get_entity_name()}_id', 'unknown')
+        return f'{self._get_entity_name().title()} {entity_id}: {self.previous_state} → {self.state}'
```
curious why include previous_state if this is just the BaseState, and not a transition right?

bmartel replied:
States keep the previous state for quick access without additional queries to know what the previous state was. (This is so that flows which can move to multiple possible next states, we can know at a glance which the previous was).

bmartel replied:
This is not a transition, just a denormalized optimization.

hakan458 commented on label_studio/fsm/models.py:
```diff
@@ -0,0 +1,168 @@
+    def _get_entity_name(self) -> str:
+        """Extract entity name from model name (e.g., TaskState → task)"""
+        model_name = self.__class__.__name__
+        if model_name.endswith('State'):
```
When adding states to the platform will you / should we be following this pattern always?

bmartel replied:
Yes, this is a great callout! The preferred method is convention over configuration and to follow as it documents to name your states according to the models they anchor to. If the model is an inherited proxy model which would extend the state choices, then you override this in method and return the string key that is intended. (You will see this in PRs for the model states and choices implementation).

hakan458 commented on label_studio/fsm/models.py:
```diff
@@ -0,0 +1,168 @@
+        """Get all states since a specific timestamp"""
+        entity_field = f'{cls._get_entity_field_name()}'
+        queryset = cls.objects.filter(**{entity_field: entity})
+        return UUID7Field.filter_since_time(queryset, since).order_by('id')
```
Just want to say the usage of UUID7 is amazing and makes things so much easier / more performant 👏🏼 great choice

bmartel replied:
This was a suggestion from @wesleylima 🚀

hakan458 commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,420 @@
+        """
+        entity_key = entity_name.lower()
+
+        if entity_key in self._models:
```
when/why would we overwrite an existing state model? In what scope will this persist? E.g. can I have a different state model in the app vs rqworker for `task` ?

bmartel replied:
No this is strictly an extension mechanism to allow for LSO -> LSE model states, as they differ greatly.

hakan458 commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,420 @@
+    def __init__(self):
+        self._transitions: Dict[str, Dict[str, 'BaseTransition']] = {}
+
+    def register(self, entity_name: str, transition_name: str, transition_class: 'BaseTransition'):
```
just curious does the system allow to register multiple transitions, with the same current state and target state? would that be good practice? Im sure you have thought of some examples

bmartel replied:
Currently no, but that is a really neat idea 🤔. Will have to think more on that, but generally speaking I believe it could be made to work like that. Might be a more declarative way to go about the multi option transitions.

hakan458 replied:
How is it enforced then, that we cannot register multiple transitions with the same state?

hakan458 commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,422 @@
+        # Query database using state model registry
+        state_model = get_state_model_for_entity(entity)
+        if not state_model:
+            logger.warning(f'No state model found for {entity._meta.model_name}')
```
Maybe add something like "when attempting to get current state". Otherwise unclear where this is coming from

hakan458 commented on label_studio/fsm/state_manager.py:
```diff
@@ -0,0 +1,422 @@
+        """
+        state_model = get_state_model_for_entity(entity)
+        if not state_model:
+            raise StateManagerError(f'No state model found for {entity._meta.model_name}')
```
+ " ... when transitioning state" or similar

hakan458 commented on label_studio/fsm/README.md:
```diff
@@ -0,0 +1,354 @@
+
+# Bulk operations for performance
+orders = Order.objects.all()[:1000]
+states = StateManager.bulk_get_current_states(orders)
```
does this exist in LSO? Or only LSE?

bmartel replied:
LSE, I'll remove this from here.

bmartel committed:
updating docs

bmartel committed:
removing unused code

bmartel committed:
applying feedback from CR

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
structure the logging so it can be made useful for metrics in grafana

bmartel committed:
fix implementation

bmartel committed:
Merge branch 'develop' into fb-fit-587-fsm

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17496376340)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17497319557)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

wesleylima commented on label_studio/fsm/registry.py:
```diff
@@ -0,0 +1,390 @@
+        transition = transition_class(**transition_data)
+
+        # Get current state information
+        from fsm.state_manager import StateManager
```
Is it useful to have these import ins the function call?

bmartel replied:
Checking this, we seem to have an import cycle between state_manager and this file due to this method, I'll take note and see if this can get refactored so we don't require this. At this point I believe it was the easiest way to have the functionality culminate.

bmartel replied:
I'll take a stab at refactoring this and update the tests accordingly. Nice catch!

bmartel committed:
refactor code to remove function imports

bmartel committed:
refactor code to remove redundant code

bmartel committed:
updating docs

bmartel committed:
fixing lint errors
